drivers/i2s/i2s_mcux_flexcomm.c

```diff
@@ -0,0 +1,965 @@
+{
+	void *buffer;
+
+	while (k_msgq_get(&stream->in_queue, &buffer, K_NO_WAIT) == 0) {
+		k_mem_slab_free(mem_slab, &buffer);
+	}
+	while (k_msgq_get(&stream->out_queue, &buffer, K_NO_WAIT) == 0) {
+		k_mem_slab_free(mem_slab, &buffer);
+	}
+}
+
+static void i2s_mcux_tx_stream_disable(const struct device *dev, bool drop)
+{
+	const struct i2s_mcux_config *cfg = dev->config;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->tx;
+	I2S_Type *base = cfg->base;
+
+	LOG_DBG("Stopping DMA channel %u for TX stream", stream->channel);
+	dma_stop(stream->dev_dma, stream->channel);
+
+	/* Clear TX error interrupt flag */
+	base->FIFOSTAT = I2S_FIFOSTAT_TXERR(1U);
+	I2S_DisableInterrupts(base, (uint32_t)kI2S_TxErrorFlag);
+
+	if (base->CFG1 & I2S_CFG1_MAINENABLE_MASK) {
+		/* Wait until all transmitted data get out of FIFO */
+		while ((base->FIFOSTAT & I2S_FIFOSTAT_TXEMPTY_MASK) == 0U) {
+		}
+		/*
+		 * The last piece of valid data can be still being transmitted from
+		 * I2S at this moment
+		 */
+		/* Write additional data to FIFO */
+		base->FIFOWR = 0U;
+		while ((base->FIFOSTAT & I2S_FIFOSTAT_TXEMPTY_MASK) == 0U) {
+		}
+
+		/* At this moment the additional data is out of FIFO, we can stop I2S */
+		/* Disable TX DMA */
+		base->FIFOCFG &= (~I2S_FIFOCFG_DMATX_MASK);
+		base->FIFOCFG |= I2S_FIFOCFG_EMPTYTX_MASK;
+
+		I2S_Disable(base);
+	}
+
+	/* purge buffers queued in the stream */
+	if (drop) {
+		i2s_purge_stream_buffers(stream, stream->cfg.mem_slab);
+	}
+}
+
+static void i2s_mcux_rx_stream_disable(const struct device *dev, bool drop)
+{
+	const struct i2s_mcux_config *cfg = dev->config;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->rx;
+	I2S_Type *base = cfg->base;
+
+	LOG_DBG("Stopping DMA channel %u for RX stream", stream->channel);
+	dma_stop(stream->dev_dma, stream->channel);
+
+	/* Clear RX error interrupt flag */
+	base->FIFOSTAT = I2S_FIFOSTAT_RXERR(1U);
+	I2S_DisableInterrupts(base, (uint32_t)kI2S_RxErrorFlag);
+
+	/* stop transfer */
+	/* Disable Rx DMA */
+	base->FIFOCFG &= (~I2S_FIFOCFG_DMARX_MASK);
+	base->FIFOCFG |= I2S_FIFOCFG_EMPTYRX_MASK;
+
+	I2S_Disable(base);
+
+	/* purge buffers queued in the stream */
+	if (drop) {
+		i2s_purge_stream_buffers(stream, stream->cfg.mem_slab);
+	}
+}
+
+static void i2s_mcux_config_dma_blocks(const struct device *dev,
+					enum i2s_dir dir, uint32_t *buffer)
+{
+	const struct i2s_mcux_config *cfg = dev->config;
+	struct i2s_mcux_data *dev_data = dev->data;
+	I2S_Type *base = cfg->base;
+	struct dma_block_config *blk_cfg;
+	struct stream *stream;
+
+	if (dir == I2S_DIR_RX) {
+		stream = &dev_data->rx;
+	} else {
+		stream = &dev_data->tx;
+	}
+
+	blk_cfg = &stream->dma_block[0];
+	memset(blk_cfg, 0, sizeof(struct dma_block_config));
+
+	if (dir == I2S_DIR_RX) {
+		uint8_t num_of_bufs;
+
+		/* Check if we have multiple RX buffers allocated */
+		num_of_bufs = k_mem_slab_num_used_get(stream->cfg.mem_slab);
```

Comment:
That looks wrong. User can start the stream, stop it and then start again without reading any data from the RX queue. The mere fact that some buffers were taken from the memory pool does not mean they were reserved for the DMA engine. It would be safer to check that `buffer[1]` is not NULL (if we ensure that it is NULL if only one DMA descriptor is used).

Reply:
This is preparing the DMA for receive. If no data is received and the user issues a STOP then all the DMA buffers will be freed in i2s_purge_stream_buffers.

Reply:
I have deleted this check as the code allows only receive memory slab with 2 buffers.